### PR TITLE
FOGL-8492: Fledge-GUI: Add flow editor feature for north service/task

### PR DESCRIPTION
**Pull request by mshariq-nerd**

### CODE REVIEW

File: src/app/components/core/north/north.module.ts [hunk 1]
```diff
@@ -34,7 +34,7 @@ const routes: Routes = [
 @NgModule({
   declarations: [
     NorthComponent,
-    AddTaskWizardComponent,
+    // AddTaskWizardComponent,
```
Comment (praveen-garg) [inline src/app/components/core/north/north.module.ts:37]: this is expected to be commented?

File: src/app/components/common/node-editor/custom-node/custom-node.component.ts [hunk 2]
```diff
@@ -83,32 +101,56 @@ export class CustomNodeComponent implements OnChanges {
...
+    if (this.data.label === 'South' || this.data.label === 'North') {
       if (this.source !== '') {
+        this.isServiceNode = true;
         this.elRef.nativeElement.style.borderColor = "#B6D7A8";
         interval(POLLING_INTERVAL)
           .pipe(takeWhile(() => this.isAlive), takeUntil(this.destroy$)) // only fires when component is alive
           .subscribe(() => {
-            this.getSouthboundServices();
+            if (this.from == 'north') {
+              this.getNorthboundTasks();
```
Comment (praveen-garg) [inline src/app/components/common/node-editor/custom-node/custom-node.component.ts:112]: Should we replace Task with Instance? Given North Instance can be a service or task.

File: src/app/components/common/node-editor/node-editor.component.html [hunk 3]
```diff
@@ -1,97 +1,100 @@
...
-        <button class="button is-small" (click)="closeModal('delete-service-dialog')">Cancel</button>
-        <button class="button is-small is-danger" (click)="deleteService()"
-            [appDisableUntilResponse]="reenableButton">Delete</button>
-    </footer>
+  <header class="modal-card-head">
+    <span class="modal-card-title is-size-6">Delete Service</span>
+    <button class="delete" aria-label="close" (click)="closeModal('delete-service-dialog')"></button>
+  </header>
+  <section class="modal-card-body">
+    Deleting this service can not be undone. Continue?
```
Comment (praveen-garg) [inline src/app/components/common/node-editor/node-editor.component.html:78]: replace this with instance Name. Can also be task? 

cannot

File: src/app/components/common/node-editor/node-editor.component.html [hunk 4]
```diff
@@ -1,97 +1,100 @@
...
-        <button class="button is-small" (click)="closeModal('delete-filter-dialog')">Cancel</button>
-        <button class="button is-small is-danger" (click)="deleteFilterFromPipeline()"
-            [appDisableUntilResponse]="reenableButton">Delete</button>
-    </footer>
+  <header class="modal-card-head">
+    <span class="modal-card-title is-size-6">Delete Filter</span>
+    <button class="delete" aria-label="close" (click)="closeModal('delete-filter-dialog')"></button>
+  </header>
+  <section class="modal-card-body">
+    Deleting this filter can not be undone. Continue?
```
Comment (praveen-garg) [inline src/app/components/common/node-editor/node-editor.component.html:93]: As above

File: src/app/components/common/node-editor/north.ts [hunk 5]
```diff
@@ -0,0 +1,25 @@
+import { ClassicPreset } from "rete";
+
+
+export class North extends ClassicPreset.Node {
+  height = 115;
```
Comment (praveen-garg) [inline src/app/components/common/node-editor/north.ts:5]: 110

File: src/app/components/common/node-editor/south.ts [hunk 6]
```diff
@@ -2,25 +2,24 @@ import { ClassicPreset } from "rete";
 
 
 export class South extends ClassicPreset.Node {
-    height = 115;
-    width = 220;
-    parent?: string;
+  height = 115;
```
Comment (praveen-garg) [inline src/app/components/common/node-editor/south.ts:5]: 110

File: src/app/components/core/north/add-task-wizard/add-task-wizard.component.ts [hunk 7]
```diff
@@ -261,7 +274,16 @@ export class AddTaskWizardComponent implements OnInit, OnDestroy {
             const name = payload.name;
             this.uploadScript(name, files);
           }
-          this.router.navigate(['/north']);
+          if (this.source === 'flowEditor') {
+            this.ngProgress.start();
+            setTimeout(() => {
+              this.router.navigate(['flow/north', response['name']])
+              this.ngProgress.done();
+            }, 3000);
```
Comment (praveen-garg) [inline src/app/components/core/north/add-task-wizard/add-task-wizard.component.ts:282]: We should add inline comment for this 3s timeout.

File: src/app/components/core/north/add-task-wizard/add-task-wizard.component.ts [hunk 8]
```diff
@@ -305,7 +327,16 @@ export class AddTaskWizardComponent implements OnInit, OnDestroy {
             const name = payload.name
             this.uploadScript(name, files);
           }
-          this.router.navigate(['/north']);
+          if (this.source === 'flowEditor') {
+            this.ngProgress.start();
+            setTimeout(() => {
+              this.router.navigate(['/flow/editor/north', response['name'], 'details']);
+              this.ngProgress.done();
+            }, 3000);
```
Comment (praveen-garg) [inline src/app/components/core/north/add-task-wizard/add-task-wizard.component.ts:335]: As above. Also we can test in slow (RPi0 with lab script) setup

File: src/app/components/core/north/add-task-wizard/add-task-wizard.component.ts [hunk 9]
```diff
@@ -305,7 +327,16 @@ export class AddTaskWizardComponent implements OnInit, OnDestroy {
             const name = payload.name
             this.uploadScript(name, files);
           }
-          this.router.navigate(['/north']);
+          if (this.source === 'flowEditor') {
+            this.ngProgress.start();
+            setTimeout(() => {
```
Comment (hemantgarg923) [inline src/app/components/core/north/add-task-wizard/add-task-wizard.component.ts:332]: This setTimeout could be removed now as we have handled API response with fork join on node editor component

File: src/app/components/core/south/add-service-wizard/add-service-wizard.component.ts [hunk 10]
```diff
@@ -94,7 +95,12 @@ export class AddServiceWizardComponent implements OnInit, OnDestroy {
     const last = <HTMLElement>document.getElementsByClassName('step-item is-active')[0];
     const id = last.getAttribute('id');
     if (+id === 1) {
-      this.router.navigate(['/south']);
+      console.log('move previous');
+      if (this.source) {
```
Comment (hemantgarg923) [inline src/app/components/core/south/add-service-wizard/add-service-wizard.component.ts:99]: This check could also be removed as both are redirecting to same page

File: src/app/components/core/south/add-service-wizard/add-service-wizard.component.ts [hunk 11]
```diff
@@ -256,14 +262,14 @@ export class AddServiceWizardComponent implements OnInit, OnDestroy {
             const name = payload.name
             this.uploadScript(name, files);
           }
-          if(this.source === 'flowEditor'){
+          if (this.source === 'flowEditor') {
             this.ngProgress.start();
             setTimeout(() => {
-              this.router.navigate(['/south/flow'], { queryParams: { source: response['name'] } });
+              this.router.navigate(['/flow/editor/south', response['name'], 'details']);
```
Comment (hemantgarg923) [inline src/app/components/core/south/add-service-wizard/add-service-wizard.component.ts:268]: Similarly this setTimeout could also be removed

File: src/app/components/core/north/add-task-wizard/add-task-wizard.component.ts [hunk 12]
```diff
@@ -261,7 +274,16 @@ export class AddTaskWizardComponent implements OnInit, OnDestroy {
             const name = payload.name;
             this.uploadScript(name, files);
           }
-          this.router.navigate(['/north']);
+          if (this.source === 'flowEditor') {
+            this.ngProgress.start();
+            setTimeout(() => {
```
Comment (hemantgarg923) [inline src/app/components/core/north/add-task-wizard/add-task-wizard.component.ts:279]: Similarly this setTimeout could also be removed

File: src/app/components/core/north/add-task-wizard/add-task-wizard.component.ts [hunk 13]
```diff
@@ -89,7 +98,11 @@ export class AddTaskWizardComponent implements OnInit, OnDestroy {
     const last = <HTMLElement>document.getElementsByClassName('step-item is-active')[0];
     const id = last.getAttribute('id');
     if (+id === 1) {
-      this.router.navigate(['/north']);
+      if (this.source) {
```
Comment (hemantgarg923) [inline src/app/components/core/north/add-task-wizard/add-task-wizard.component.ts:101]: This check could be removed as both paths are redirecting to same page

File: src/app/components/core/north/north.module.ts (hunk 1 above)
Comment (mshariq-nerd) [reply src/app/components/core/north/north.module.ts]: yes

File: src/app/components/common/node-editor/node-editor.component.html [hunk 14]
```diff
@@ -1,98 +1,103 @@
...
-                        Add Filter
-                    </a>
-                </li>
-            </ul>
-        </nav>
-        <ng-container *ngIf="source && rolesService.hasEditPermissions() && !isAddFilterWizard">
-            <button class="button is-small is-link" (click)="save()">Save</button>
-        </ng-container>
+  <div class="navbar-item pb-0">
+    <nav class="breadcrumb mb-0" aria-label="breadcrumbs">
```
Comment (praveen-garg) [inline src/app/components/common/node-editor/node-editor.component.html:3]: make it mb-2

File: src/app/components/core/south/south.module.ts [hunk 15]
```diff
@@ -38,7 +34,7 @@ const routes: Routes = [
 @NgModule({
   declarations: [
     SouthComponent,
-    AddServiceWizardComponent,
+    // AddServiceWizardComponent,
```
Comment (praveen-garg) [inline src/app/components/core/south/south.module.ts:37]: remove?

**Review by praveen-garg — COMMENTED** [review]
Role based testing done?

Viewer especially

**Comment by praveen-garg** [comment]
I have added as a task, Delete alert says service, We can make it either a check based or simply say instance?

When added as a task, Show in tiny light gray text `Task` next to instance name in the node.

When hover in node, context menu icon appears white, should have something similar to background but lighter.

Don't show -1 when sent count is -1.

Test Ctrl/Cmd+Z behaviour for undo?!

File: src/app/components/common/node-editor/custom-node/custom-node.component.ts [hunk 16]
```diff
@@ -301,6 +333,34 @@ export class CustomNodeComponent implements OnChanges {
...
+          this.service.status = this.fetchedTask?.status;
+          let readingCount = this.fetchedTask.sent;
+          this.service.readingCount = readingCount;
+          this.service.schedule_enabled = this.fetchedTask.enabled;
+          if (this.service.schedule_enabled === true) {
+            this.isEnabled = true;
+          }
+          else {
+            this.isEnabled = false;
+          }
```
Comment (monika-Dianomic) [inline src/app/components/common/node-editor/custom-node/custom-node.component.ts:352]: Can't we write it as `this.isEnabled = this.service.schedule_enabled;`?

File: src/app/components/common/node-editor/custom-node/custom-node.component.ts [hunk 17]
```diff
@@ -315,8 +375,8 @@ export class CustomNodeComponent implements OnChanges {
           }, 0)
           this.service.assetCount = assetCount;
           this.service.readingCount = readingCount;
-          this.service.schedule_enabled = String(this.fetchedService.schedule_enabled);
-          if (this.service.schedule_enabled === 'true') {
+          this.service.schedule_enabled = this.fetchedService.schedule_enabled;
+          if (this.service.schedule_enabled === true) {
```
Comment (monika-Dianomic) [inline src/app/components/common/node-editor/custom-node/custom-node.component.ts:379]: same as above

File: src/app/components/core/south/south.component.ts [hunk 18]
```diff
@@ -144,9 +144,14 @@ export class SouthComponent implements OnInit, OnDestroy {
   }
 
   addSouthService() {
+    console.log('add service');
```
Comment (monika-Dianomic) [inline src/app/components/core/south/south.component.ts:147]: remove?

File: src/app/components/core/south/add-service-wizard/add-service-wizard.component.ts [hunk 19]
```diff
@@ -94,7 +95,12 @@ export class AddServiceWizardComponent implements OnInit, OnDestroy {
     const last = <HTMLElement>document.getElementsByClassName('step-item is-active')[0];
     const id = last.getAttribute('id');
     if (+id === 1) {
-      this.router.navigate(['/south']);
+      console.log('move previous');
```
Comment (monika-Dianomic) [inline src/app/components/core/south/add-service-wizard/add-service-wizard.component.ts:98]: remove?

File: src/app/components/common/node-editor/nodes/north.ts [hunk 20]
```diff
@@ -0,0 +1,32 @@
...
+  NameControl,
+  PluginControl,
+  StatusControl
+} from "../controls/common-custom-control";
+import { SentReadingsControl } from "../controls/north-custom-control";
+
+export class North extends ClassicPreset.Node {
+  height = 110;
+  width = 220;
+  parent?: string;
```
Comment (monika-Dianomic) [inline src/app/components/common/node-editor/nodes/north.ts:13]: These lines are being used repeatedly. Can't we fetch it from any common place?

File: src/app/components/common/node-editor/custom-node/custom-node.component.ts [hunk 21]
```diff
@@ -249,16 +281,16 @@ export class CustomNodeComponent implements OnChanges {
   }
```
Comment (monika-Dianomic) [inline src/app/components/common/node-editor/custom-node/custom-node.component.ts:281]: Handling for "North" is missing here.

**Review by monika-Dianomic — APPROVED** [review]

**Review by hemantgarg923 — APPROVED** [review]

**Pull request merged**